### DISCUSSION TITLE
Call service.AddKakfa() multiple times

### DISCUSSION THREAD

**Discussion by ravi-kotha** [Farfetch/kafkaflow#157] [Q&A] [answered]
Hi All,

Is there a  way we can invoke service.AddKafka() multiple times with the same app?
I am trying to build abstractions for various producer contexts and in some cases I will have to use more then one of these abstractions in a single application.

I tried to call this multiple times and looks like it does not resolve the instances properly after that, anyway I can achieve this?

**Comment by filipeesch** [comment, marked as answer]
Hi @ravi-kotha,

You can call AddKafka multiple times if your application has more than one ServiceCollection/ServiceProvider, but is not usual.

Maybe you can achieve that by calling AddKafka just one time and call the AddProducer multiple times, like:

```csharp
services.AddKafkaFlowHostedService(kafka => kafka
    .AddCluster(cluster => cluster
    {
        foreach(var producerCOnfig in producersConfig)
            cluster.AddProducer(...)
    })
);
```

**Reply by ravi-kotha** [reply]
I am already adding multiple producers, but these will live in separate nugets, So I  have a Web App which will consume these different packages for producing different stuff to different topics and I wanted to abstract all the DI into the nugets, but I realized if I did that in the nugets then I cannot add both the nugets in the web app at the same time since they both call the AddKafka piece separately. Both these nugets have extension methods which will use the consuming apps ServiceCollection object.

**Reply by filipeesch** [reply]
Wouldn't be better to create extension methods on `IClusterConfigurationBuilder` in your packages with the producer configurations? And use like this:

```csharp
services.AddKafkaFlowHostedService(kafka => kafka
    .AddCluster(cluster => cluster
    {
            cluster.AddProducerOne(...) // your extension method
            cluster.AddProducerTwo(...) // your extension method
    })
);
```

**Reply by ravi-kotha** [reply]
That could work, let me try. Thanks.

**Reply by ravi-kotha** [reply]
You are the man 👏👏, that works. Thank you so much for the quick replies.
